pmd-core/src/main/java/net/sourceforge/pmd/util/BaseResultProducingCloseable.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.util;
+
+import java.util.function.Consumer;
+
+/**
+ * Base class for an autocloseable that produce a result once it has
+ * been closed.
+ *
+ * @param <T> Type of the result
+ */
+public abstract class BaseResultProducingCloseable<T> implements AutoCloseable {
+
+    private boolean closed;
+
+    protected final void ensureOpen() {
+        if (closed) {
+            throw new IllegalStateException("Listener closed");
+        }
+    }
+
+    /**
+     * Returns the result.
+     *
+     * @throws IllegalStateException If this instance has not been closed yet
+     */
+    public final T getResult() {
+        if (!closed) {
+            throw new IllegalStateException("Cannot get result before listener is closed");
+        }
+
+        return getResultImpl();
+    }
+
+    /** Produce the final result. */
+    protected abstract T getResultImpl();
+
+    /**
+     * @implNote Call super
```

Comment:
this sound a bit error prone. How about providing a protected void closeImpl() method and making close final?

Reply:
You're right this would be cleaner